src/libraries/System.Linq/src/System/Linq/PartialArrayEnumerator.cs

```diff
@@ -0,0 +1,42 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Linq
+{
+    /// <summary>Enumerator for iterating through part of an array.</summary>
+    internal sealed class PartialArrayEnumerator<T> : IEnumerator<T>
+    {
+        private readonly T[] _array;
+        private readonly int _count;
+        private int _index = -1;
+
+        public PartialArrayEnumerator(T[] array, int count)
+        {
+            Debug.Assert(array is not null);
+            _array = array;
+            _count = count;
+        }
+
+        public bool MoveNext()
+        {
+            if (_index + 1 < _count)
```

Comment:
```suggestion
            if (_index < _count - 1)
```
* to avoid potential overflow
* `_count` isn't used elsewhere, so `_count - 1` could be stored in the field instead of just `_count`, thus avoiding that repeated calc

Reply:
> to avoid potential overflow

Overflow isn't possible. _count is at most Array.MaxLength, which is < int.MaxValue.

> that repeated calc

It's not actually repeated. The JIT does a good job of avoiding the duplicate increment (though not when moving the +1 to be -1 on the other side).

[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhgMBAOwwMA+pACu0gNxce/YaPFSZsqQBMYCVdR4M13eg2AQIAGwYBZCEJgA5IxnIAKAJQWOALMBADMGbz1JQwQGAGoGcgYAHjlFaX9TMx5OTKysyOjY2JM8vOIAdgYMKAUYEtKAXxogngqGEOw7XDqAptzzfqsbeycXd09SPwCc0okwiIMjZNSIJQwM2ZnZngKjIvrtpkrq2oOsvpbLSo6unv6+swCh2wdnVw8EDDop/q280PCuxiKXkq2kDDgCQ2pT+2yB+0uWTaJzuswu/TMbRu3TOfQaQA==)

Reply:
Cool that the JIT does this.

I thought of something like [MoveNext4](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHuGAHKAJYA3bBhgMBAOwwMA+pACu0gNxce/YaPFSZ8iEowBZKQtzlV1Hn0EixE6XKkATGAgtW13eg2AQIAGwZDCCEYADlXDHIACgBKTw4EqwEAMwZo2WdXBgBqBnIGAB45RWl4yyseTgrKyszJFwQcnPdayuIAdgYMKAUYVtqAXxoknk6GFOx/XH6E4ZqGBO9fAKCQ8MiUOITqtok0jKyEIpL9aWNJU3Jyvd29nnrG5oH78Z6+l6t5vdGvLsnprMFvNBkA==), but as you wrote there's no need to change.